### PR TITLE
Ref #77: Custom widget support for all types added to ui:widget.

### PR DESCRIPTION
**Pull request by n1k0**
### Custom field components

You can provide your own field components to a uiSchema for basically any json schema data type, by specifying a `ui:field` property.

For example, let's create and register a dumb `geo` component handling a *latitude* and a *longitude*:

```jsx
const schema = {
  type: "object",
  required: ["lat", "lon"],
  properties: {
    lat: {type: "number"},
    lon: {type: "number"}
  }
};

// Define a custom component for handling the root position object
class GeoPosition extends React.Component {
  constructor(props) {
    super(props);
    this.state = {...props.formData};
  }

  onChange(name) {
    return (event) => {
      this.setState({
        [name]: parseFloat(event.target.value)
      }, () => this.props.onChange(this.state));
    };
  }

  render() {
    const {lat, lon} = this.state;
    return (
      <div>
        <input type="number" value={lat} onChange={this.onChange("lat")} />
        <input type="number" value={lon} onChange={this.onChange("lon")} />
      </div>
    );
  }
}

// Define the custom field component to use for the root object
const uiSchema = {"ui:field": "geo"};

// Define the custom field components to register; here our "geo"
// custom field component
const fields = {geo: GeoPosition};

// Render the form with all the properties we just defined passed
// as props
render(<Form
  schema={schema}
  uiSchema={uiSchema}
  fields={fields}/>);
```

Note: Registered fields can be reused accross the entire schema.

Feedback? @chmac
r=?@magopian

### CODE REVIEW

**Comment by n1k0** [comment]
Hmm this patch is wrong as we need to distinguish between fields and widgets; providing a replacement for an object field is definitely not the same as providing an input (widget) for eg. a string.

I might introduce a new `ui:field` to disambiguate the two distinct use cases.

**Comment by n1k0** [comment]
This is now ready for early review. r=? @magopian

**Comment by almet** [comment]
How do you do the correlation between "geo" and the component used here?

File: README.md [hunk 1]
```diff
@@ -271,7 +272,12 @@ render(
 
 ## Advanced customization
 
-### Custom widgets
+The API allows to specify your own custom *widgets* and *fields* components:
+
+- A *widget* represents a HTML tag for the user to enter data, eg. `input`, `select`, etc.
```
Comment (almet) [inline README.md]: nit: should this be "an HTML tag" ?

File: README.md [hunk 2]
```diff
@@ -271,7 +272,12 @@ render(
...
+- A *field* usually wraps one or more widgets and most often handle internal field state; think of a field as a form row, including the labels.
```
Comment (almet) [inline README.md]: nit: handles

File: README.md [hunk 3]
```diff
@@ -342,6 +348,57 @@ The following props are passed to the widget component:
...
+
+class GeoPosition extends Component {
+  constructor(props) {
+    super(props);
+    this.state = {...props.formData};
+  }
+
+  onChange(name) {
+    return (event) => {
+      this.setState({
```
Comment (almet) [inline README.md]: It might be a react idiom, but I don't understand this line. It might be useful to add some comments explaining why we need two argumens here and why we're passing only a partial state to `setState`.

File: playground/app.js [hunk 4]
```diff
@@ -28,6 +28,41 @@ const cmOptions = {
...
+  constructor(props) {
+    super(props);
+    this.state = {...props.formData};
+  }
+
+  onChange(name) {
+    return (event) => {
+      this.setState({
+        [name]: parseFloat(event.target.value)
+      }, () => this.props.onChange(this.state));
```
Comment (almet) [inline playground/app.js]: I don't understand why we need this?

Comment (almet) [inline playground/app.js]: I don't understand why we need this?

File: playground/app.js [hunk 5]
```diff
@@ -176,6 +211,7 @@ class App extends Component {
               formData={this.state.formData}
               onChange={data => this.setState({formData: data.formData})}
               onSubmit={data => this.setState({formData: data.formData})}
+              fields={{geo: GeoPosition}}
```
Comment (almet) [inline playground/app.js]: So that's where the connection is made. The documentation should reflect that to be easier to grasp.

File: src/components/Form.js [hunk 6]
```diff
@@ -119,12 +125,11 @@ if (process.env.NODE_ENV !== "production") {
     schema: PropTypes.object.isRequired,
     uiSchema: PropTypes.object,
     formData: PropTypes.any,
+    fields: PropTypes.object,
```
Comment (almet) [inline src/components/Form.js]: Good move :)

File: src/components/fields/SchemaField.js [hunk 7]
```diff
@@ -105,7 +115,10 @@ if (process.env.NODE_ENV !== "production") {
   SchemaField.propTypes = {
     schema: PropTypes.object.isRequired,
     uiSchema: PropTypes.object,
-    registry: PropTypes.object,
+    registry: PropTypes.shape({
+      widgets: PropTypes.object.isRequired,
+      fields: PropTypes.object.isRequired,
```
Comment (almet) [inline src/components/fields/SchemaField.js]: Should we also check the shape of the `fields` object?

**Comment by almet** [comment]
This is great! thanks for doing this, I think it will be a really useful addition to the library. Everything looks good to my non-react expert eyes.

**Comment by n1k0** [comment]
> How do you do the correlation between "geo" and the component used here?

With this line from the docs sample:

```js
const uiSchema = {"ui:field": "geo"};
```

File: README.md (hunk 3 above)
Comment (n1k0) [inline README.md]: Two lines below we're calling `this.onChange("lat")`, which returns a function, which is our event handler for this specific field. I wonder what kind of comment is needed here.

File: playground/app.js (hunk 4 above)
Comment (n1k0) [inline playground/app.js]: Why we need what?

File: playground/app.js (hunk 5 above)
Comment (n1k0) [inline playground/app.js]: Well, that's part of the example from the doc, and furthermore the registration is approached in the section before... Well, I'll see if I can improve it then.

File: src/components/fields/SchemaField.js (hunk 7 above)
Comment (n1k0) [inline src/components/fields/SchemaField.js]: No because it's not predictable (or at least PropTypes don't expose any API for ensuring a map is composed from a given typed data structure).

File: playground/app.js (hunk 5 above)
Comment (n1k0) [inline playground/app.js]: I'm realizing the README wasn't updated accordingly to this. Will fix.

File: src/components/fields/SchemaField.js (hunk 7 above)
Comment (n1k0) [inline src/components/fields/SchemaField.js]: Of course, I lied:

>  // An object with property values of a certain type
>  optionalObjectOf: React.PropTypes.objectOf(React.PropTypes.number),

Comment (n1k0) [inline src/components/fields/SchemaField.js]: Addressed in 856a1b47fa15b404d10fd197ccf0f5207301a99a

File: README.md (hunk 3 above)
Comment (almet) [inline README.md]: Previous comment was about the `this.setState` call, which is taking two arguments: I don't understand what the second argument is for.

File: playground/app.js (hunk 4 above)
Comment (almet) [inline playground/app.js]: Oh, ok, I think I got it. We first set the state and then we call `props.onChange`, probably to propagate the action.

I was mistaken by:

- The `setState` which isn't promise-based.
- The fact that we're calling `onChange` on the props, probably to cascade the changes.

File: playground/app.js (hunk 5 above)
Comment (almet) [inline playground/app.js]: Thanks!

File: README.md (hunk 3 above)
Comment (n1k0) [inline README.md]: Oh, that callback is to ensure the next operation is performed *after* the new state has been taken into account and the call to render() being executed. That's important so we avoid races where the state is declared to be a and we propagate b.

File: playground/app.js (hunk 4 above)
Comment (n1k0) [inline playground/app.js]: :)

**Comment by chmac** [comment]
@n1k0 Wow, I missed all the excitement. The change looks **very** cool. Don't have sufficient react / es6 experience to talk as to the code quality, so can only comment on the API.

On first pass, the `uiSchema` -> `fields` was not immediately obvious to me. Now I've read it again, I see the logic, and I like it. I can define a field once, and use it multiple times. I could even define the `fields` map globally and pass it to all my forms, even if all the field components are not necessarily used in every form I guess.

It would be great to add a list of the props to the docs. For example, the `this.state = {...props.formData}` line doesn't make a lot of sense unless you know what `formData` is. Something similar to the list in the widgets component would be ideal.

Bottom line, thanks a lot @n1k0, this change is a huge benefit for us. I think this may now be the react form package I've been dreaming of... :-)
